_posts/2017-08-02-July.md

```diff
@@ -0,0 +1,45 @@
+---
+layout: post
+title: "Dev Diary (July 2017)"
+author: "Corey Montella"
+tags: []
+---
+
+We've had quite a busy July! Let's get to it:
+
+## v0.4 milestones
+
+In July we started using v0.4 to build internal projects, in order to find bugs in the runtime and test the limits of its performance. If you've hit performance walls with the v0.3 runtime, you'll be happy to know the Rust runtime is several orders of magnitude (10x - 100x) faster. Beyond the runtime itself, we've started fleshing out the tooling:
+
+- Server - The sever works like it did in v0.2, hosting an Eve instance listening on a websocket. We also include a static file server that hosts the output of your program.
+- Browser Client - The browser client connects to a remote Eve server to which it streams records like key presses; and receives records like ticks of a timer. These changes are rendered to the browser by watchers, which extend Eve's functionality into a foreign language. So far we have interfaces for writing watchers in Rust and Typescript.
+
+- Persistance - This is one of the features that has been missing for a while and frequently requested. Persistence is implemented by writing to an append-only binary file at a user specified location. Right now this is a flag passed to the Eve executable, but this could be controlled from within Eve itself. Transactions are written to this log as they are executed in the runtime. On load, the log of transactions is replayed, which restores the runtime to the state it was at when it was terminated. There are some obvious limitations to this when we start talking about Eve in a distributed context, but this at least gets us a starting place.
+
+What about crashes?
+
+- Library - The entire standard library has to be re-implemented in Rust. This is largely a matter of just the work of wrapping native Rust functions, but we want to hold off on that until we implement a custom number system we have planned (more on that in a future post). We've added old watchers like file, console, and compiler; as well as new watchers like http and json. These newest additions allow you to parse JSON files into records, or send HTTP requests to API endpoints and parse the response.
```

Comment:
Doesn't have to be (we can interface with JS watchers still) but we're choosing to for a few reasons:
1) performance benefits
2) JS ones require a browser context (so can't be used in terminal only)
3) Taking the time to do a general cleanup/improvement pass